New evaluators config (#722)
langchain-ai/langsmith-docs: docs/evaluation/how_to_guides/llm_as_judge.mdx

import {
  CodeTabs,
  python,
  typescript,
} from "@site/src/components/InstructionsWithCode";

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import '@site/src/css/custom.css';

# How to define an LLM-as-a-judge evaluator

:::info Key concepts
- [LLM-as-a-judge evaluator](/evaluation/concepts#llm-as-judge)
:::

LLM applications can be challenging to evaluate since they often generate conversational text with no single correct answer.

> This guide shows you how to define a LLM-as-a-judge evaluator for [offline evaluation](/evaluation/concepts#offline-evaluation) using either the LangSmith SDK or the UI. Note: To run evaluations in real-time on your production traces, refer to [setting up online evaluations](/observability/how_to_guides/online_evaluations#configure-llm-as-judge-evaluators). 

Reviewer: ubernit: is it "an LLM-as-a..."? i know its a consonant but phonetically 'a llm' sounds weird, think you say "an llm" when speaking

> <Tabs className="interface-tabs">

Reviewer: the tabs look kinda funky, imo better to just have separate headers

Author: updating styling slightly but def could use improvement

content becomes quite long if we dont tab it. since its two very diff group of users i think we should keep content separated for easier readability

  <TabItem value="sdk" label="SDK">

<div id="sdk-content">
## Pre-built evaluators

> Pre-built evaluators are a useful starting point setting up evaluations. Refer to [pre-built evaluators](./prebuilt_evaluators) for how to use pre-built evaluators with LangSmith.

Reviewer: ```suggestion
Pre-built evaluators are a useful starting point for setting up evaluations. Refer to [pre-built evaluators](./prebuilt_evaluators) for how to use pre-built evaluators with LangSmith.
```

## Create your own LLM-as-a-judge evaluator

> For complete control of evaluator logic, create your own LLM-as-a-judge evaluator and run it using the LangSmith SDK ([Python](https://docs.smith.langchain.com/reference/python/reference) / [TypeScript](https://docs.smith.langchain.com/reference/js)).

Reviewer: not sure how valuable it is to just link to the api ref home page

Author: i always need to go searching for the links so my pref is to link to them whenever we mention + link to specific methods when guides are specific enough

  typescript,
} from "@site/src/components/InstructionsWithCode";

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# How to define an LLM-as-a-judge evaluator

:::info Key concepts

- [LLM-as-a-judge evaluator](/evaluation/concepts#llm-as-judge)

:::

LLM applications can be difficult to evaluate because often they're generating conversational text for which there's no single "right" answer.

> A valuable way to evaluate such applications is to use a second LLM to judge the outputs of the first.

Author: ya confusing. gonna remove

/>

See [here](./custom_evaluator) for more on how to write a custom evaluator.
  </TabItem >
  <TabItem value="ui" label="UI">

  ## Pre-built evaluators
  
  Pre-built evaluators are a useful starting point when setting up evaluations. The LangSmith UI supports the following pre-built evaluators:
  

>   - **Hallucination**: Detect factually incorrect outputs.

Reviewer 2: minor nit: I would mention this is with regards to a reference.

/>

See [here](./custom_evaluator) for more on how to write a custom evaluator.
</div>

  </TabItem >
  <TabItem value="ui" label="UI">
  <div id="ui-content">

  ## Pre-built evaluators
  
  Pre-built evaluators are a useful starting point when setting up evaluations. The LangSmith UI supports the following pre-built evaluators:
  
  - **Hallucination**: Detect factually incorrect outputs.
  - **Correctness**: Check semantic similarity to a reference.
  - **Conciseness**: Evaluate whether an answer is a concise response to a question. 
  - **Code checker**: Verify correctness of code answers.
  
  You can configure these evaluators::
  - When running an evaluation using the [playground](/prompt_engineering/concepts#prompt-playground)
  - As part of a dataset to [automatically run experiments over a dataset](/evaluation/how_to_guides/bind_evaluator_to_dataset)
  - When running an [online evaluation](/observability/how_to_guides/online_evaluations#configure-llm-as-judge-evaluators)
  

>   Pre-built evaluators easy to customize. By default, pre-built evaluators assume that you are running the evaluation over the full input, output and (if applicable) reference output. To specify which keys to run the evaluation over, or to change the LLM-as-a-judge prompt you can edit the evaluator. See more details below. 
>   
>   ## Customize your LLM-as-a-judge evaluator

Reviewer: ```suggestion
  ## Customize your LLM-as-a-judge evaluator

  By default, pre-built evaluators assume that you are running the evaluation over the full input, output and (if applicable) reference output. If your application and dataset call for specific instructions or formatting, you can customize the default prompt and which parts of the input/output/reference output should be passed to the judge LLM. You can also configure the judge LLM.
```

>   Pre-built evaluators easy to customize. By default, pre-built evaluators assume that you are running the evaluation over the full input, output and (if applicable) reference output. To specify which keys to run the evaluation over, or to change the LLM-as-a-judge prompt you can edit the evaluator. See more details below. 
>   
>   ## Customize your LLM-as-a-judge evaluator

Reviewer: ```suggestion
  ## Customize your LLM-as-a-judge evaluator

  By default, pre-built evaluators assume that you are running the evaluation over the full input, output and (if applicable) reference output. You can customize the default prompt and which parts of the input/output/reference output should be passed to the judge LLM.
  
```

  
  ![](./static/playground_evaluator.gif)
  
 ### Select/create the evaluator
  - In the playground or from a dataset: Select the **+Evaluator** button
  - From a tracing project: Select **Add rules**, configure your rule and select **Apply evaluator**
  
  Select the **Create your own evaluator option**. Alternativley, you may start by selecting a pre-built evaluator and editing it. 
  
### Configure the evaluator

> Edit the prompt, a model, set of criteria to evaluate the on, and set a feedback schema for the evaluator. 

Reviewer: ```suggestion
Edit the prompt, judge LLM, and evaluation criteria.
```

#### Prompt

Create a new prompt, or choose an existing prompt from the [prompt hub](/prompt_engineering/quickstarts/quickstart_ui).

- **Create your own prompt**: Create a custom prompt inline. 

- **Pull a prompt from the prompt hub**: Use the **Select a prompt** dropdown to select from an existing prompt. You can't edit these prompts directly within the prompt editor, but you can view the prompt and the schema it uses. To make changes, edit the prompt in the playground and commit the version, and then pull in your new prompt in the evaluator.

#### Model

Select the desired model from the provided options.

#### Mapping variables

> Use variable mapping to indicate the variables that are passed into your evaluator prompt from your run or example. To aid with variable mapping, a sample example (or run) is provided for reference. Use the dropdown to select the parts of the run that are relevant for the evalaution.

Reviewer: ```suggestion
Click on the the variables in your prompt and use the dropdown to map them to the relevant parts of the input, output, or reference output. A sample run and (if applicable) dataset example are displayed for reference.
```

> You may edit/remove variables as needed. For example if you are evaluating a metric such as conciseness, you typically don't need a reference output so you may remove that variable. 

Reviewer: ```suggestion
To add prompt variables type the variable with double curly brackets `{{prompt_var}}` if using mustache formatting (the default) or single curly brackets `{prompt_var}` if using f-string formatting. To delete a prompt variable use the backspace key.
```

#### Preview

> Previewing the prompt will show you an example of what the formatted prompt will look like. This preview pulls a sample input, output into the prompt. In the case of a evaluator configured as part of a dataset or in the playground, the preview will also pull reference output from an example in your dataset.

Reviewer: ```suggestion
Previewing the prompt will show you of what the formatted prompt will look like using the reference run and dataset example shown on the right.
```

> #### Improve your evaluator with few-shot examples

Reviewer: think this should be in the other how to guide (or at least copied over there)

Author: moved to other guide + linked ✅

  
  If you're using few-shot examples, LangSmith has a feature in beta that allows the most relevant examples to be dynamically selected at runtime using [dynamic few-shot examples](/evaluation/how_to_guides/index_datasets_for_dynamic_few_shot_example_selection). 
</details>

#### Feedback configuration

Feedback configuration is the scoring criteria that your LLM-as-a-judge evaluator will use. Think of this as the rubric that your evaluator will grade based on. Scores will be added as [feedback](/observability/concepts#feedback) to a run or example. Defining feedback for your evaluator:

1. **Name the feedback key**: This is the name that will appear when viewing evaluation results. Names should be unique across experiments. 

2. **Add a description**: Describe what the feedback represents. 

> 3. **Choose feedback type**: 

Reviewer: ```suggestion
3. **Choose a feedback type**: 
```

- **Boolean**: True/false feedback.
- **Categorical**: Select from predefined categories. 
- **Continuous**: Numerical scoring within a specified range.

> Feedback is added as output schema to your LLM-as-a-judge evaluator prompt. If you are pulling in a prompt from the hub, you must add feedback configuration as part of the output schema for your prompt in the playground first.

Reviewer: ```suggestion
If you are using an existing prompt from the hub, you must add an output schema to the prompt before configuring an evaluator to use it. Each top-level key in the output schema will be treated as a separate piece of feedback.
```
